File: src/main/java/AoC2022_09.java
import static com.github.pareronia.aoc.IntegerSequence.Range.range;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toList;
import static java.util.stream.Collectors.toSet;

import java.util.Arrays;
import java.util.List;
import java.util.Set;

import com.github.pareronia.aoc.geometry.Direction;
import com.github.pareronia.aoc.geometry.Position;
import com.github.pareronia.aoc.geometry.Vector;
import com.github.pareronia.aocd.Aocd;
import com.github.pareronia.aocd.Puzzle;

public class AoC2022_09 extends AoCBase {
    
    private final List<Direction> moves;
    
    private AoC2022_09(final List<String> input, final boolean debug) {
        super(debug);
        this.moves = input.stream()
            .flatMap(line -> {
                final String[] splits = line.split(" ");
                final Direction[] h = new Direction[Integer.parseInt(splits[1])];
                Arrays.fill(h, Direction.fromString(splits[0]));
                return Arrays.stream(h);
            })
            .collect(toList());
    }
    
    public static final AoC2022_09 create(final List<String> input) {
        return new AoC2022_09(input, false);
    }

    public static final AoC2022_09 createDebug(final List<String> input) {
        return new AoC2022_09(input, true);
    }
    
    void printRope(final Position[] rope) {
        log(Arrays.stream(rope)
            .map(r -> String.format("(%d,%d)", r.getX(), r.getY()))
            .collect(joining(" ")));
    }
    
    private Position catchup(final Position head, final Position tail) {
        final int dx = head.getX() - tail.getX();
        final int dy = head.getY() - tail.getY();
        if (Math.abs(dx) > 1 || Math.abs(dy) > 1) {
            final var move = Vector.of(
                dx < 0 ? - 1 : dx > 0 ? 1 : 0,
                dy < 0 ? - 1 : dy > 0 ? 1 : 0);
            return tail.translate(move);
        }
        return tail;
    }

    private void moveRope(final Position[] rope, final Direction move) {
        rope[0] = rope[0].translate(move);
        range(1, rope.length, 1).forEach(j ->
                rope[j] = catchup(rope[j - 1], rope[j]));
        printRope(rope);
    }
    
    private int solve(final int size) {
        final Position[] rope = new Position[size];
        Arrays.fill(rope, Position.of(0, 0));
        final Set<Position> seen = this.moves.stream()
            .map(move -> {
                moveRope(rope, move);
                return rope[rope.length - 1];
            })
            .collect(toSet());
        log(seen.size());
        log(seen);
        return seen.size();
    }
    
    @Override
    public Integer solvePart1() {
        return solve(2);
    }

    @Override
    public Integer solvePart2() {
        return solve(10);
    }

    public static void main(final String[] args) throws Exception {
        assert AoC2022_09.createDebug(TEST1).solvePart1() == 13;
        assert AoC2022_09.createDebug(TEST1).solvePart2() == 1;
        assert AoC2022_09.createDebug(TEST2).solvePart2() == 36;

        final Puzzle puzzle = Aocd.puzzle(2022, 9);
        final List<String> inputData = puzzle.getInputData();
        puzzle.check(
            () -> lap("Part 1", AoC2022_09.create(inputData)::solvePart1),
            () -> lap("Part 2", AoC2022_09.create(inputData)::solvePart2)
        );
    }

    private static final List<String> TEST1 = splitLines("""
        R 4
        U 4
        L 3
        D 1
        R 4
        D 1
        L 5
        R 2
        """);
    private static final List<String> TEST2 = splitLines("""
        R 5
        U 8
        L 8
        D 3
        R 17
        D 10
        L 25
        U 20
        """);
}

File: src/main/python/AoC2022_09.py
#! /usr/bin/env python3
#
# Advent of Code 2022 Day 9
#


import itertools
import sys

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

TEST1 = """\
R 4
U 4
L 3
D 1
R 4
D 1
L 5
R 2
"""
TEST2 = """\
R 5
U 8
L 8
D 3
R 17
D 10
L 25
U 20
"""

MOVES = {"U": (0, 1), "D": (0, -1), "L": (-1, 0), "R": (1, 0)}

Position = tuple[int, int]
Move = tuple[int, int]
Input = tuple[Move, ...]
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return tuple(
            MOVES[m]
            for move, amount in (line.split() for line in input_data)
            for m in itertools.repeat(move, int(amount))
        )

    def solve(self, moves: tuple[Move, ...], size: int) -> int:
        def _move_rope(rope: list[Position], move: Move) -> list[Position]:
            def _catchup(head: Position, tail: Position) -> Position:
                dx = head[0] - tail[0]
                dy = head[1] - tail[1]
                if abs(dx) > 1 or abs(dy) > 1:
                    return (
                        tail[0] + (-1 if dx < 0 else 1 if dx > 0 else 0),
                        tail[1] + (-1 if dy < 0 else 1 if dy > 0 else 0),
                    )
                return tail

            rope[0] = (rope[0][0] + move[0], rope[0][1] + move[1])
            for i in range(1, len(rope)):
                rope[i] = _catchup(rope[i - 1], rope[i])
            return rope

        rope = [(0, 0)] * size
        seen = {(rope := _move_rope(rope, move))[-1] for move in moves}
        return len(seen)

    def part_1(self, moves: Input) -> Output1:
        return self.solve(moves, size=2)

    def part_2(self, moves: Input) -> Output2:
        return self.solve(moves, size=10)

    @aoc_samples(
        (
            ("part_1", TEST1, 13),
            ("part_2", TEST1, 1),
            ("part_2", TEST2, 36),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2022, 9)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
